lib/ansible/modules/cloud/docker/docker_swarm.py

```diff
@@ -0,0 +1,403 @@
+
+description:
+     - Init a new Swarm cluster.
+     - Add/Remove nodes or managers to an existing cluster.
+options:
+    advertise_addr:
+        description:
+            - Externally reachable address advertised to other nodes.
+            - This can either be an address/port combination
+                in the form ``192.168.1.1:4567``, or an interface followed by a
+                port number, like ``eth0:4567``. If the port number is omitted,
+                the port number from the listen address is used. If
+                ``advertise_addr`` is not specified, it will be automatically
+                detected when possible.
+    listen_addr:
+        description:
+            - Listen address used for inter-manager communication.
+            - This can either be an address/port combination in the form
+                ``192.168.1.1:4567``,
+                or an interface followed by a port number, like ``eth0:4567``.
+                If the port number is omitted, the default swarm listening port
+                is used.
+        default: 0.0.0.0:2377
+    force:
+        description:
+            - Use with state C(init) to force creating a new Swarm, even if already part of one.
+            - Use with state C(leave) to Leave the swarm even if this node is a manager.
+        type: bool
+        default: false
+    state:
+        description:
+            - Set to C(init), to create a new cluster.
+            - Set to C(join), to join an existing cluster.
+            - Set to C(leave), to leave an existing cluster.
+            - Set to C(remove), to remove a node.
+        required: true
+        default: join
+        choices:
+          - init
+          - join
+          - leave
+          - remove
+    node_id:
+        description:
+            - Used with I(state=remove). Swarm id of the node to remove.
+    join_token:
+        description:
+            - Used with I(state=join). Swarm token used to join a swarm cluster.
+    remote_addrs:
+        description:
+            - Used with I(state=join). Remote address of a manager to connect to.
+    task_history_retention_limit:
+        description:
+            - Maximum number of tasks history stored.
+    snapshot_interval:
+        description: Number of logs entries between snapshot.
+    keep_old_snapshots:
+        description: Number of snapshots to keep beyond the current snapshot.
+    log_entries_for_slow_followers:
+        description: Number of log entries to keep around to sync up slow followers after a snapshot is created.
+    heartbeat_tick:
+        description: Amount of ticks (in seconds) between each heartbeat.
+    election_tick:
+        description: Amount of ticks (in seconds) needed without a leader to trigger a new election.
+    dispatcher_heartbeat_period:
+        description: The delay for an agent to send a heartbeat to the dispatcher.
+    node_cert_expiry:
+        description: Automatic expiry for nodes certificates.
```

Comment:
So I noticed this long list of options without defaults or choices. Possibly this is exactly what is intended.
However, in a lot of cases there is an implicit default that could be mentioned (even when it's not enforced when missing).

If you initialize a new swarm I expect there are defaults set.

Reply:
Done

Reply:
What are the default values for te below parameters if you do not specify the parameter on creation ?
- task_history_retention_limit
- keep_old_snapshots
- log_entries_for_slow_followers
- heartbeat_tick
- election_tick
- dispatcher_heartbeat_period
- node_cert_expiry
- ca_force_rotate
- autolock_managers

Reply:
In the API, all these parameters are optional and by default value is None. They are set only if value is not None. So it's just a docker default value. How do yo suggest to do that ? It's not a default ansible value. By default these parameters are not set and not push to docker API.

Reply:
Add the default value to the description. The default values should be documented, if not, create a new swarm and check its value.

How is the module idempotent in this way ? If you would modify a value, how would it know that task made a change or not ?